### ISSUE TITLE
[CMake] generate compile_commands.json during cmake stage

### ISSUE THREAD

**Issue by eranif** [wxWidgets/wxWidgets#23094]
The file `compile_commands.json` is used by `clangd` ( **the** code completion engine for almost all decent C++ IDEs these days)
This file contains the build instructions for how to compile each file in the source tree.

`CMake` can generate it by adding a single line to the top level `CMakeLists.txt` file:

```diff
diff --git a/CMakeLists.txt b/CMakeLists.txt
index ee8fd239c1..2233b766cb 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -8,6 +8,7 @@
 #############################################################################

 cmake_minimum_required(VERSION 3.0)
+set(CMAKE_EXPORT_COMPILE_COMMANDS 1)

 if(NOT CMAKE_CONFIGURATION_TYPES)
     get_property(HAVE_MULTI_CONFIG_GENERATOR GLOBAL PROPERTY GENERATOR_IS_MULTI_CONFIG)

```

By adding this, the code completion works like a charm

Thanks

**Comment by vadz** [comment]
Can't this be done by just using `-DCMAKE_EXPORT_COMPILE_COMMANDS=1` on the command line? If so, I'm not sure it makes much sense to hardcode it here.

**Comment by imciner2** [comment]
IMO, I don't think this should be hardcoded in the build system, since most users won't actually want this if they just want to build the library to use it. It is better to just pass `-DCMAKE_EXPORT_COMPILE_COMMANDS=ON` on the CMake command line when you want to use clangd.

**Comment by eranif** [comment]
I disagree here. Most users don't really know how VScode or any other IDE that supports `LSP` (Language Server Protocol) works. Creating this file by default will make it "just work" without googling and banging their heads :D 

I don't see why not adding this line. It does no harm, it has 0 impact on the user (it does not slow the cmake stage as well)
I can't find a good reason why NOT adding it

**Comment by vadz** [comment]
If CMake doesn't do it by default, there must be some reason for it... And it's still much simpler to enable it than to disable it if you do _not_ want it for some reason (e.g. because you have `compile_commands.json` coming from some other place, maybe generated by BEAR).

**Comment by eranif** [comment]
Usually, when people are building with `CMake` they build out of the source tree (to avoid polluting the source tree)
So they (all the developer I have worked with are using this method):

```bash
mkdir build-dir
cd $_
cmae .. -DCMAKE_BUILD_TYPE=BlaBla ...
make -j$(nproc)
```
The `compile_commands.json` is created inside the build directory, so not really an issue.
Anyways, up to you

I am good with either
